framework/include/kernels/MatBodyForce.h

```diff
@@ -0,0 +1,63 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "BodyForce.h"
+#include "JvarMapInterface.h"
+#include "DerivativeMaterialInterface.h"
+
+// Forward Declarations
+class Function;
+
+/**
+ * This kernel creates a body force that is modified by a mask defined
+ * as a material. Common uses of this would be to turn off or change the
+ * body force in certain regions of the mesh.
+ */
+template <bool is_ad, class Parent>
+class MatBodyForceTempl : public Parent
+{
+public:
+  static InputParameters validParams();
+
+  MatBodyForceTempl(const InputParameters & parameters);
+
+protected:
+  virtual GenericReal<is_ad> computeQpResidual() override;
+
+  const GenericMaterialProperty<Real, is_ad> & _property;
+
+  /// name of the nonlinear variable (needed to retrieve the derivative material properties)
```

Comment:
Same here, should this be in the nonad constructor?

Reply:
I'll push a fix